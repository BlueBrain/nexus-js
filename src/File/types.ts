import { Context } from '../Resource/types';
import { HttpConfigTypes } from '../utils/http';

export interface NexusFileResponse {
  '@context': Context;
  '@id': string;
  '@type': 'File';
  _bytes: number;
  _digest: {
    _algorithm: string;
    _value: string;
  };
  _filename: string;
  _mediaType: string;
  _self: string;
  _constrainedBy: string;
  _project: string;
  _rev: number;
  _deprecated: boolean;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
}

export interface FetchFileOptions extends FetchRawFileOptions {
  shouldFetchFile?: boolean;
}

export interface FetchRawFileOptions {
  receiveAs?: HttpConfigTypes;
}

export interface CreateFileOptions {
  storage?: string;
  [key: string]: any;
}
