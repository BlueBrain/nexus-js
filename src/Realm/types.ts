import { PaginationSettings, PaginatedList } from '../utils/types';
import Realm from '.';

export interface RealmUtils {
  get(realmLabel: string, rev?: number): Promise<Realm>;
  list(listRealmOptions?: ListRealmOptions): Promise<PaginatedList<Realm>>;
  create(realmLabel: string, realmPayload: CreateRealmPayload): Promise<Realm>;
  update(
    realmLabel: string,
    rev: number,
    realmPayload: CreateRealmPayload,
  ): Promise<Realm>;
  deprecate(realmLabel: string, rev: number): Promise<Realm>;
}

export type Context = string | string[];

export interface RealmResponseCommon {
  '@id': string;
  '@type': string;
  name: string;
  openidConfig: string;
  _label: string;
  _issuer: string;
  _authorizationEndpoint: string;
  _tokenEndpoint: string;
  _userInfoEndpoint: string;
  _endSessionEndpoint: string;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
  _deprecated: boolean;
}
export interface RealmResponse extends RealmResponseCommon {
  '@context': Context;
}
export interface ListRealmResponse {
  '@context': Context;
  _total: number;
  _results: RealmResponseCommon[];
}

export interface ListRealmOptions {
  from?: PaginationSettings['from'];
  size?: PaginationSettings['size'];
  deprecated?: boolean;
  [key: string]: any;
}

export interface CreateRealmPayload {
  name: string;
  openIdConfig: string;
  logo?: string;
}
