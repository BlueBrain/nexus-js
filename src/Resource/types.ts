export interface ResourceResponseCommon {
  '@id': string;
  '@type'?: string | string[];
  _self: string;
  _constrainedBy: string;
  _project: string;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
  _deprecated: boolean;
  // This is for the rest of the dataset's data
  [key: string]: any;
}

export interface ListResourceResponse {
  '@context'?: Context;
  _total: number;
  _results: ResourceResponseCommon[];
}

export interface ResourceResponse extends ResourceResponseCommon {
  '@context'?: Context;
}

export type Context = string | object;

export interface ListResourceResponseError {
  '@context': Context;
  code: string;
  message: string;
}

export interface ResourceResponseError {
  '@context': Context;
  code: string;
  message: string;
  ref: string;
}

export interface ResourceListTagResponse {
  '@context': Context;
  tags: string[];
}

export interface CreateResourcePayload {
  resourceId?: string;
  type?: string[];
  context: { [field: string]: string };
  [field: string]: any;
}

export interface UpdateResourcePayload {
  type?: string[];
  context?: { [field: string]: string };
  [field: string]: any;
}

export interface ListResourceOptions {
  full_text_search?: string;
  from?: number;
  size?: number;
  deprecated?: boolean;
  [key: string]: any;
}

export enum ResourceGetFormat {
  JSON_LD = 'application/ld+json',
  DOT = 'text/vnd.graphviz',
  N_TRIPLES = 'application/ntriples',
}

export type ResourceGetFormats =
  | ResourceGetFormat.DOT
  | ResourceGetFormat.JSON_LD
  | ResourceGetFormat.N_TRIPLES;

export interface GetResourceOptions {
  expanded: boolean;
}
