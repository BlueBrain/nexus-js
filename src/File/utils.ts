// @ts-ignore
import FormData from 'isomorphic-form-data';
import { NexusFileResponse } from './types';
import { httpPost, HttpConfigTypes, httpGet } from '../utils/http';
import NexusFile from './index';
import { isBrowser } from '../utils';
import { ReadStream } from 'fs';
import { Readable } from 'stream';

export async function createFile(
  orgLabel: string,
  projectLabel: string,
  file: File | Blob | ReadableStream | ReadStream | Readable,
): Promise<NexusFile> {
  try {
    // create new form data
    const form = new FormData();
    // add file
    form.append('file', file);
    // if in Node.js, we need to manually set headers
    let extraHeaders = {};
    if (!isBrowser) {
      extraHeaders = form.getHeaders();
    }
    const fileResponse: NexusFileResponse = await httpPost(
      `/files/${orgLabel}/${projectLabel}`,
      form,
      {
        extraHeaders,
        sendAs: HttpConfigTypes.FILE,
      },
    );
    return new NexusFile(orgLabel, projectLabel, fileResponse);
  } catch (error) {
    throw error;
  }
}

export async function getFileSelf(
  selfUrl: string,
  shouldFetchFile: boolean = false,
): Promise<NexusFile> {
  const fileResponse: NexusFileResponse = await httpGet(selfUrl, {
    useBase: false,
  });
  const [id, schema, projectLabel, orgLabel, ...rest] = selfUrl
    .split('/')
    .reverse();
  const file = new NexusFile(orgLabel, projectLabel, fileResponse);
  if (shouldFetchFile) {
    await file.getFile();
  }
  return file;
}

export async function getFile(
  orgLabel: string,
  projectLabel: string,
  fileId: string,
  shouldFetchFile: boolean = false,
): Promise<NexusFile> {
  const fileResponse: NexusFileResponse = await httpGet(
    `/files/${orgLabel}/${projectLabel}/${fileId}`,
  );
  const file = new NexusFile(orgLabel, projectLabel, fileResponse);
  if (shouldFetchFile) {
    await file.getFile();
  }
  return file;
}

export async function getRawFile(
  selfURL: string,
): Promise<string | Blob | ReadStream | ArrayBuffer> {
  const rawFile: string | Blob | ReadStream | ArrayBuffer = await httpGet(
    selfURL,
    {
      useBase: false,
      extraHeaders: {
        Accept: '*/*',
      },
      receiveAs: HttpConfigTypes.BASE64,
    },
  );
  return rawFile;
}
