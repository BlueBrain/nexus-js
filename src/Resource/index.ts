import {
  Context,
  ResourceResponse,
  ResourceGetFormats,
  GetResourceOptions,
  ResourceLink,
  ExpandedResource,
} from './types';
import {
  getResource,
  getSelfResource,
  getSelfResourceRawAs,
  updateResource,
  updateSelfResource,
  listResources,
  createResource,
  tagResource,
  deprecateResource,
  deprecateSelfResource,
  tagSelfResource,
  listTags,
  listSelfTags,
  getIncomingLinks,
  getOutgoingLinks,
} from './utils';
import { PaginatedList, PaginationSettings } from '../utils/types';

export const DEFAULT_GET_RESOURCE_OPTIONS: GetResourceOptions = {
  expanded: false,
};

export const RESOURCE_METADATA_KEYS = [
  '@context',
  '@id',
  '@type',
  '_self',
  '_constrainedBy',
  '_project',
  '_createdAt',
  '_createdBy',
  '_updatedAt',
  '_updatedBy',
  '_rev',
  '_deprecated',
];

export default class Resource<T = {}> {
  readonly orgLabel: string;
  readonly projectLabel: string;
  readonly context?: Context;
  readonly type?: string[];
  readonly self: string;
  readonly id: string;
  readonly constrainedBy: string;
  readonly project: string;
  readonly createdAt: string;
  readonly createdBy: string;
  readonly updatedAt: string;
  readonly updatedBy: string;
  readonly rev: number;
  readonly deprecated: boolean;
  readonly data: T;
  readonly raw: ResourceResponse;
  readonly resourceURL: string;
  expanded?: ExpandedResource;

  static getSelf = getSelfResource;
  static getSelfRawAs = getSelfResourceRawAs;
  static getIncomingLinks = getIncomingLinks;
  static getOutgoingLinks = getOutgoingLinks;
  static get = getResource;
  static list = listResources;
  static create = createResource;
  static updateSelf = updateSelfResource;
  static update = updateResource;
  static deprecate = deprecateResource;
  static deprecateSelf = deprecateSelfResource;
  static tag = tagResource;
  static tagSelf = tagSelfResource;
  static listTags = listTags;
  static listSelfTags = listSelfTags;
  static formatName(raw: ResourceResponse): string {
    const name =
      raw['skos:prefLabel'] ||
      raw['rdfs:label'] ||
      raw['schema:name'] ||
      raw['label'] ||
      raw['name'] ||
      // TODO move to File Class ?
      raw['_filename'] ||
      raw['@id'];
    return name;
  }

  constructor(
    orgLabel: string,
    projectLabel: string,
    resourceResponse: ResourceResponse,
  ) {
    this.raw = resourceResponse;
    this.orgLabel = orgLabel;
    this.projectLabel = projectLabel;
    this.id = resourceResponse['@id'];
    this.context = resourceResponse['@context'];
    this.self = resourceResponse._self;
    this.constrainedBy = resourceResponse._constrainedBy;
    this.project = resourceResponse._project;
    this.createdAt = resourceResponse._createdAt;
    this.createdBy = resourceResponse._createdBy;
    this.updatedAt = resourceResponse._updatedAt;
    this.updatedBy = resourceResponse._updatedBy;
    this.rev = resourceResponse._rev;
    this.deprecated = resourceResponse._deprecated;
    // make type an array of sting, even if we only get a single string
    if (resourceResponse['@type']) {
      if (Array.isArray(resourceResponse['@type'])) {
        this.type = resourceResponse['@type'] as string[];
      } else {
        this.type = [resourceResponse['@type']] as string[];
      }
    }
    // Put user custom fields in "data" key
    this.data = Object.keys(resourceResponse).reduce(
      (memo: any, key: string) => {
        if (!RESOURCE_METADATA_KEYS.includes(key)) {
          memo[key] = resourceResponse[key];
        }
        return memo;
      },
      {},
    );
    this.resourceURL = `/resources/${this.orgLabel}/${
      this.projectLabel
    }/_/${encodeURIComponent(this.id)}`;
  }

  async getAs(format: ResourceGetFormats): Promise<any> {
    return await getSelfResourceRawAs(this.self, format);
  }

  get name(): string {
    return Resource.formatName(this.raw);
  }

  async update({
    context,
    ...data
  }: {
    context?: Context;
    [field: string]: any;
  }): Promise<Resource> {
    return Resource.updateSelf(
      this.resourceURL,
      this.rev,
      {
        context,
        ...data,
      },
      this.orgLabel,
      this.projectLabel,
    );
  }

  async getExpanded() {
    this.expanded = await getSelfResourceRawAs(`${this.self}?format=expanded`);
    return this.expanded;
  }

  async getIncomingLinks(
    paginationSettings: PaginationSettings,
  ): Promise<PaginatedList<ResourceLink>> {
    let expandedID;
    if (this.expanded && this.expanded['@id']) {
      expandedID = this.expanded['@id'];
    } else {
      await this.getExpanded();
      expandedID = (this.expanded as ExpandedResource)['@id'];
    }
    return await getIncomingLinks(
      this.orgLabel,
      this.projectLabel,
      expandedID,
      paginationSettings,
    );
  }

  async getOutgoingLinks(
    paginationSettings: PaginationSettings,
  ): Promise<PaginatedList<ResourceLink>> {
    let expandedID;
    if (this.expanded && this.expanded['@id']) {
      expandedID = this.expanded['@id'];
    } else {
      await this.getExpanded();
      expandedID = (this.expanded as ExpandedResource)['@id'];
    }
    return await getOutgoingLinks(
      this.orgLabel,
      this.projectLabel,
      expandedID,
      paginationSettings,
    );
  }

  toJSON() {
    return {
      ...Object.getOwnPropertyNames(this).reduce((memo: any, key) => {
        memo[key] = (this as any)[key];
        return memo;
      }, {}),
      name: this.name,
    };
  }
}
