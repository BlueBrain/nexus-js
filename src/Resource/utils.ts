import { PaginatedList, Project } from '..';
import Resource, { DEFAULT_GET_RESOURCE_OPTIONS } from '.';
import {
  httpGet,
  httpPut,
  httpPost,
  httpDelete,
  HttpConfigTypes,
} from '../utils/http';
import {
  CreateResourcePayload,
  ResourceListTagResponse,
  ListResourceOptions,
  ResourceResponse,
  ListResourceResponse,
  ResourceResponseCommon,
  Context,
  ResourceGetFormat,
  ResourceGetFormats,
  GetResourceOptions,
  ResourceLink,
} from './types';
import { buildQueryParams } from '../utils';
import { PaginationSettings } from '../utils/types';
import { SparqlViewQueryResponse } from '../View/SparqlView/types';
import { getSparqlView } from '../View/utils';

// Fetch a resource as raw data for any number of
// content negotiation formats,
// e.g. JSON-LD, nTriples, and DOT
export async function getSelfResourceRawAs(
  selfUrl: string,
  format: ResourceGetFormats = ResourceGetFormat.JSON_LD,
): Promise<any> {
  try {
    return await httpGet(selfUrl, {
      useBase: false,
      extraHeaders: {
        Accept: format,
      },
      receiveAs:
        format === ResourceGetFormat.JSON_LD
          ? HttpConfigTypes.JSON
          : HttpConfigTypes.TEXT,
    });
  } catch (error) {
    throw error;
  }
}

export async function getSelfResource(
  selfUrl: string,
  getResourceOptions: GetResourceOptions = DEFAULT_GET_RESOURCE_OPTIONS,
): Promise<Resource> {
  try {
    const [id, schema, projectLabel, orgLabel, ...rest] = selfUrl
      .split('/')
      .reverse();
    const url = `${selfUrl}${
      getResourceOptions.expanded ? '?format=expanded' : ''
    }`;
    const resourceResponse: ResourceResponse = await getSelfResourceRawAs(url);
    const resource = new Resource(
      orgLabel,
      projectLabel,
      resourceResponse,
      getResourceOptions,
    );
    return resource;
  } catch (error) {
    throw error;
  }
}

export async function getResource(
  orgLabel: string,
  projectLabel: string,
  schemaId: string,
  resourceId: string,
  getResourceOptions: GetResourceOptions = DEFAULT_GET_RESOURCE_OPTIONS,
): Promise<Resource> {
  const projectResourceURL = `/resources/${orgLabel}/${projectLabel}/${schemaId}/${resourceId}${
    getResourceOptions.expanded ? '?format=expanded' : ''
  }`;
  try {
    const resourceResponse: ResourceResponse = await httpGet(
      projectResourceURL,
    );
    const resource = new Resource(
      orgLabel,
      projectLabel,
      resourceResponse,
      getResourceOptions,
    );
    return resource;
  } catch (error) {
    throw error;
  }
}

export async function listResources(
  orgLabel: string,
  projectLabel: string,
  options?: ListResourceOptions,
): Promise<PaginatedList<Resource>> {
  try {
    const opts = buildQueryParams(options);
    const listResourceResponses: ListResourceResponse = await httpGet(
      `/resources/${orgLabel}/${projectLabel}${opts}`,
    );
    const total: number = listResourceResponses._total;
    const index: number = (options && options.from) || 1;
    const results: Resource[] = listResourceResponses._results.map(
      (commonResponse: ResourceResponseCommon) =>
        new Resource(orgLabel, projectLabel, {
          ...commonResponse,
          '@context': listResourceResponses['@context'],
        }),
    );

    return {
      total,
      index,
      results,
    };
  } catch (error) {
    throw error;
  }
}

export async function createResource(
  orgLabel: string,
  projectLabel: string,
  schemaId: string,
  payload: CreateResourcePayload,
): Promise<Resource> {
  try {
    let resource: Resource;
    // if no resourceID provided, we post the new resource (Nexus will create the ID for us)
    if (!payload.resourceId) {
      const { context, type, ...rest } = payload;
      const resourceResponse: ResourceResponse = await httpPost(
        `/resources/${orgLabel}/${projectLabel}/${schemaId}`,
        {
          '@context': context,
          '@type': type,
          ...rest,
        },
      );
      resource = new Resource(orgLabel, projectLabel, resourceResponse);
    } else {
      const { context, type, resourceId, ...rest } = payload;
      const resourceResponse: ResourceResponse = await httpPut(
        `/resources/${orgLabel}/${projectLabel}/${schemaId}/${encodeURIComponent(
          resourceId,
        )}`,
        {
          '@context': context,
          '@type': type,
          ...rest,
        },
      );
      resource = new Resource(orgLabel, projectLabel, resourceResponse);
    }

    return resource;
  } catch (error) {
    throw error;
  }
}

export async function updateSelfResource(
  selfUrl: string,
  rev: number,
  {
    context,
    ...data
  }: {
    context?: Context;
    [field: string]: any;
  },
  orgLabel: string,
  projectLabel: string,
): Promise<Resource> {
  try {
    const resourceResponse: ResourceResponse = await httpPut(
      `${selfUrl}?rev=${rev}`,
      {
        '@context': context,
        ...data,
      },
    );
    return new Resource(orgLabel, projectLabel, resourceResponse);
  } catch (error) {
    throw error;
  }
}

export async function updateResource(
  orgLabel: string,
  projectLabel: string,
  schemaId: string,
  resourceId: string,
  rev: number,
  {
    context,
    ...data
  }: {
    context?: Context;
    [field: string]: any;
  },
): Promise<Resource> {
  const projectResourceURL = `/resources/${orgLabel}/${projectLabel}/${schemaId}/${resourceId}`;
  try {
    const resourceResponse: ResourceResponse = await httpPut(
      `${projectResourceURL}?rev=${rev}`,
      {
        '@context': context,
        ...data,
      },
    );
    return new Resource(orgLabel, projectLabel, resourceResponse);
  } catch (error) {
    throw error;
  }
}

export async function tagResource(
  orgLabel: string,
  projectLabel: string,
  schemaId: string,
  resourceId: string,
  rev: number,
  {
    tagName,
    tagFromRev,
  }: {
    tagName: string;
    tagFromRev: number;
  },
): Promise<Resource> {
  try {
    const resourceResponse: ResourceResponse = await httpPost(
      `/resources/${orgLabel}/${projectLabel}/${schemaId}/${resourceId}/tags?rev=${rev}`,
      {
        tag: tagName,
        rev: tagFromRev,
      },
    );
    return new Resource(orgLabel, projectLabel, resourceResponse);
  } catch (error) {
    throw error;
  }
}

export async function tagSelfResource(
  selfUrl: string,
  rev: number,
  {
    tagName,
    tagFromRev,
  }: {
    tagName: string;
    tagFromRev: number;
  },
  orgLabel: string,
  projectLabel: string,
): Promise<Resource> {
  try {
    const resourceResponse: ResourceResponse = await httpPost(
      `${selfUrl}/tags?rev=${rev}`,
      {
        tag: tagName,
        rev: tagFromRev,
      },
      { useBase: false },
    );
    return new Resource(orgLabel, projectLabel, resourceResponse);
  } catch (error) {
    throw error;
  }
}

export async function listTags(
  orgLabel: string,
  projectLabel: string,
  schemaId: string,
  resourceId: string,
): Promise<string[]> {
  try {
    const response: ResourceListTagResponse = await httpGet(
      `/resources/${orgLabel}/${projectLabel}/${schemaId}/${resourceId}/tags`,
    );
    return response.tags;
  } catch (error) {
    throw error;
  }
}

export async function listSelfTags(selfUrl: string): Promise<string[]> {
  try {
    const response: ResourceListTagResponse = await httpGet(`${selfUrl}/tags`, {
      useBase: false,
    });
    return response.tags;
  } catch (error) {
    throw error;
  }
}

export async function deprecateResource(
  orgLabel: string,
  projectLabel: string,
  schemaId: string,
  resourceId: string,
  rev: number,
): Promise<Resource> {
  try {
    const resourceResponse: ResourceResponse = await httpDelete(
      `/resources/${orgLabel}/${projectLabel}/${schemaId}/${resourceId}?rev=${rev}`,
    );
    return new Resource(orgLabel, projectLabel, resourceResponse);
  } catch (error) {
    throw error;
  }
}

export async function deprecateSelfResource(
  selfUrl: string,
  rev: number,
  orgLabel: string,
  projectLabel: string,
): Promise<Resource> {
  try {
    const resourceResponse: ResourceResponse = await httpDelete(
      `${selfUrl}?rev=${rev}`,
      { useBase: false },
    );
    return new Resource(orgLabel, projectLabel, resourceResponse);
  } catch (error) {
    throw error;
  }
}

export async function getOutgoingLinks(
  orgLabel: string,
  projectLabel: string,
  selfUrl: string,
  paginationSettings: PaginationSettings,
): Promise<PaginatedList<ResourceLink>> {
  return await getLinks(
    orgLabel,
    projectLabel,
    paginationSettings,
    () => `
  prefix nxv: <https://bluebrain.github.io/nexus/vocabulary/>
  SELECT ?total ?o ?p ?self

  WITH {
    SELECT DISTINCT ?p ?o ?self
        WHERE {
            graph ?g1 {
              <${selfUrl}> ?p ?o
              FILTER(isIri(?o) && ?p NOT IN (nxv:updatedBy, nxv:createdBy, nxv:constrainedBy, nxv:project, rdf:type))
            } .
            OPTIONAL {
              graph ?g2 {
              	?o nxv:constrainedBy ?aSchema .
                ?o nxv:self ?self
              }
            }
      }
} AS %resultSet


  WHERE {
     {
        SELECT (COUNT(?o) AS ?total)
        WHERE { INCLUDE %resultSet }
     }
     UNION
    {
        SELECT *
        WHERE { INCLUDE %resultSet }
        ORDER BY ?o ?p
        LIMIT ${paginationSettings.size}
        OFFSET ${paginationSettings.from}
     }
  }
  `,
  );
}

export async function getIncomingLinks(
  orgLabel: string,
  projectLabel: string,
  selfUrl: string,
  paginationSettings: PaginationSettings,
): Promise<PaginatedList<ResourceLink>> {
  return await getLinks(
    orgLabel,
    projectLabel,
    paginationSettings,
    () => `
    prefix nxv: <https://bluebrain.github.io/nexus/vocabulary/>

    SELECT ?total ?s ?p  ?self

    WITH {
      SELECT DISTINCT ?s ?p ?self
      WHERE {
        graph ?g1 {

          ?ref ?p <${selfUrl}> .
          ?s nxv:constrainedBy ?aSchema .
          ?s nxv:self ?self

        }
      }
    } AS %resultSet

    WHERE {
      {
        SELECT (COUNT(?s) AS ?total)
        WHERE { INCLUDE %resultSet }
      }
      UNION
      {
        SELECT *
        WHERE { INCLUDE %resultSet }
        ORDER BY ?s ?p
        LIMIT ${paginationSettings.size}
        OFFSET ${paginationSettings.from}
       }
      }
    `,
  );
}

export async function getLinks(
  orgLabel: string,
  projectLabel: string,
  paginationSettings: PaginationSettings,
  makeQuery: () => string,
): Promise<PaginatedList<ResourceLink>> {
  try {
    const { from } = paginationSettings;
    const view = await getSparqlView(orgLabel, projectLabel);
    const query = makeQuery();
    const paginatedResponse: SparqlViewQueryResponse = await view.query(query);
    if (paginatedResponse.results) {
      const totalBinding =
        paginatedResponse.results &&
        paginatedResponse.results.bindings.find(binding => !!binding.total);
      const queryResults = paginatedResponse.results.bindings.filter(
        binding => typeof binding['total'] === 'undefined',
      );

      const resourcePromises = queryResults.map(subjectPredicate => {
        // depending on incoming or outgoing, we'll center on subject or object
        const graphID = (subjectPredicate.s || subjectPredicate.o).value;

        // self exists here, so we can assume that
        // it is a resource in the platform
        if (!!subjectPredicate.self) {
          return Resource.getSelf(subjectPredicate.self.value);
        }

        // otherwise it is not a resource, likely DOI or URI
        return graphID;
      });
      const resources = await Promise.all(resourcePromises);
      return {
        total: totalBinding ? totalBinding.total.value : 0,
        index: from,
        results: queryResults.map((subjectPredicate, index) => {
          return {
            link: resources[index],
            isExternal: !subjectPredicate.self,
            predicate: subjectPredicate.p.value,
          };
        }),
      };
    }
    return {
      total: 0,
      index: from,
      results: [],
    };
  } catch (error) {
    throw error;
  }
}
