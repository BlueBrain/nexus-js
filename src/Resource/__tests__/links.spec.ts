import Resource from '..';
import { resetMocks, mock, mockResponses } from 'jest-fetch-mock';
import {
  mockGetByIDResourceResponse,
  mockProjectResponse,
  mockSparqlViewResponse,
  mockResourceResponse,
} from '../../__mocks__/helpers';
import { SparqlViewQueryResponse } from '../../View/SparqlView/types';
import { PaginatedList } from '../../utils/types';
import { ResourceLink } from '../types';
import { getIncomingLinks } from '../utils';

const mockIncomingLinksQueryResponse: SparqlViewQueryResponse = {
  head: { vars: ['s', 'p'] },
  results: {
    bindings: [
      {
        p: {
          type: 'uri',
          value: 'https://mynexusinstance.com/vocabs/org/project/favoriteBuddy',
        },
        s: {
          type: 'uri',
          value: 'https://mynexusinstance.com/vocabs/org/project/_/jeff',
        },
      },
    ],
  },
};

const mockIncomingLinksCountResponse: SparqlViewQueryResponse = {
  head: { vars: ['total'] },
  results: {
    bindings: [
      {
        total: {
          datatype: 'http://www.w3.org/2001/XMLSchema#integer',
          type: 'literal',
          value: '2',
        },
      },
    ],
  },
};

describe('Incoming / Outgoing Links behavior', () => {
  const resource = new Resource<{
    subject: string;
  }>('testOrg', 'testProject', mockGetByIDResourceResponse);
  describe('getLinks()', () => {});
  describe('getIncomingLinks()', () => {
    beforeEach(() => {
      mockResponses(
        [JSON.stringify(mockSparqlViewResponse)],
        [JSON.stringify(mockIncomingLinksCountResponse)],
        [JSON.stringify(mockIncomingLinksQueryResponse)],
        [JSON.stringify(mockResourceResponse)],
      );
    });

    afterEach(() => {
      resetMocks();
    });
    it('should fetch a PaginatedList of ResourceLinks using the proper SPAQRL queries', async () => {
      const links: PaginatedList<ResourceLink> = await getIncomingLinks(
        'myorg',
        'myproject',
        'http://blah.com/resourceSelfID',
        {
          from: 0,
          size: 20,
        },
      );
      const countQuery =
        'SELECT (COUNT(?s) AS ?total)  WHERE { ?s ?p <http://blah.com/resourceSelfID>}';
      const paginatedQuery =
        'SELECT ?s ?p WHERE { ?s ?p <http://blah.com/resourceSelfID>} LIMIT 20 OFFSET 0';
      expect(mock.calls[1][1].body).toEqual(countQuery);
      expect(mock.calls[2][1].body).toEqual(paginatedQuery);
      expect(links.results[0]).toHaveProperty('predicate');
      expect(links.results[0]).toHaveProperty('link');
      expect(links.results[0].link).toBeInstanceOf(Resource);
    });
  });
  describe('Resource.getOutgoingLinks()', () => {});
});
