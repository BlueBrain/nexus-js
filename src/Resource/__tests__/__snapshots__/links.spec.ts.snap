// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Incoming / Outgoing Links behavior getIncomingLinks() should fetch a PaginatedList of ResourceLinks using the proper SPAQRL queries as instance method 1`] = `
"
    prefix nxv: <https://bluebrain.github.io/nexus/vocabulary/>

    SELECT ?total ?s ?p  ?self

    WITH {
      SELECT DISTINCT ?s ?p ?self
      WHERE {
        graph ?g1 {

          ?ref ?p <https://bbp.epfl.ch/nexus/v0/data/bbp/morphology/reconstructedcell/v0.1.0/29d3a491-d3b7-49b6-9033-99017513a8ae> .
          ?s nxv:constrainedBy ?aSchema .
          ?s nxv:self ?self

        }
      }
    } AS %resultSet

    WHERE {
      {
        SELECT (COUNT(?s) AS ?total)
        WHERE { INCLUDE %resultSet }
      }
      UNION
      {
        SELECT *
        WHERE { INCLUDE %resultSet }
        ORDER BY ?s ?p
        LIMIT 20
        OFFSET 0
       }
      }
    "
`;

exports[`Incoming / Outgoing Links behavior getIncomingLinks() should work as a static method 1`] = `
"
    prefix nxv: <https://bluebrain.github.io/nexus/vocabulary/>

    SELECT ?total ?s ?p  ?self

    WITH {
      SELECT DISTINCT ?s ?p ?self
      WHERE {
        graph ?g1 {

          ?ref ?p <http://blah.com/resourceSelfID> .
          ?s nxv:constrainedBy ?aSchema .
          ?s nxv:self ?self

        }
      }
    } AS %resultSet

    WHERE {
      {
        SELECT (COUNT(?s) AS ?total)
        WHERE { INCLUDE %resultSet }
      }
      UNION
      {
        SELECT *
        WHERE { INCLUDE %resultSet }
        ORDER BY ?s ?p
        LIMIT 20
        OFFSET 0
       }
      }
    "
`;

exports[`Incoming / Outgoing Links behavior getOutgoingLinks() should fetch a PaginatedList of ResourceLinks using the proper SPAQRL queries as instance method 1`] = `
"
  prefix nxv: <https://bluebrain.github.io/nexus/vocabulary/>
  SELECT ?total ?o ?p ?self

  WITH {
    SELECT DISTINCT ?p ?o ?self
        WHERE {
            graph <https://bbp.epfl.ch/nexus/v0/data/bbp/morphology/reconstructedcell/v0.1.0/29d3a491-d3b7-49b6-9033-99017513a8ae/graph> {
              ?s ?p ?o
              FILTER(isIri(?o) && ?p NOT IN (nxv:updatedBy, nxv:createdBy, nxv:constrainedBy, nxv:project, rdf:type, nxv:self))
            } .
            OPTIONAL {
              graph ?g2 {
              	?o nxv:constrainedBy ?aSchema .
                ?o nxv:self ?self
                FILTER(?g2 != <https://bbp.epfl.ch/nexus/v0/data/bbp/morphology/reconstructedcell/v0.1.0/29d3a491-d3b7-49b6-9033-99017513a8ae/graph>)
              }
            }
      }
} AS %resultSet


  WHERE {
     {
        SELECT (COUNT(?o) AS ?total)
        WHERE { INCLUDE %resultSet }
     }
     UNION
    {
        SELECT *
        WHERE { INCLUDE %resultSet }
        ORDER BY ?o ?p
        LIMIT 20
        OFFSET 0
     }
  }
  "
`;

exports[`Incoming / Outgoing Links behavior getOutgoingLinks() should work just as well as a static method 1`] = `
"
  prefix nxv: <https://bluebrain.github.io/nexus/vocabulary/>
  SELECT ?total ?o ?p ?self

  WITH {
    SELECT DISTINCT ?p ?o ?self
        WHERE {
            graph <http://blah.com/resourceSelfID/graph> {
              ?s ?p ?o
              FILTER(isIri(?o) && ?p NOT IN (nxv:updatedBy, nxv:createdBy, nxv:constrainedBy, nxv:project, rdf:type, nxv:self))
            } .
            OPTIONAL {
              graph ?g2 {
              	?o nxv:constrainedBy ?aSchema .
                ?o nxv:self ?self
                FILTER(?g2 != <http://blah.com/resourceSelfID/graph>)
              }
            }
      }
} AS %resultSet


  WHERE {
     {
        SELECT (COUNT(?o) AS ?total)
        WHERE { INCLUDE %resultSet }
     }
     UNION
    {
        SELECT *
        WHERE { INCLUDE %resultSet }
        ORDER BY ?o ?p
        LIMIT 20
        OFFSET 0
     }
  }
  "
`;
